README.txt

Polyphony v. 1.5.5 (2010-03-04)
=================================

What is Polyphony?
------------------
Polyphony is a set of user-interface components to go along with the Harmoni
Framework. These components include a Wizard System, Repository access actions,
log-browsing actions, user-management forms, and a number of other user interface
components that may be useful for web-based software that is built on Harmoni.

Usage of Polyphony is not required for Harmoni usage, but most of the Polyphony
components rely on Harmoni.

Polyphony is currently used in the curricular applications built by the Curricular
Technologies group at Middlebury College: Segue 2 and Concerto.

About Harmoni:

The Harmoni Project is an effort led by the Curricular Technologies group at
Middlebury College to build an application framework and standards-based
infrastructure bus to support the rapid development and easy maintenance of
curricular it projects. The project is built entirely using PHP's OOP (Object
Oriented Programming) model, allowing the framework code to be easily extended and enhanced.

At the core of the Harmoni Application Framework is an implementation of The Open
Knowledge Initiative's (O.K.I) Open Service Interface Definitions (OSIDs). The OSIDs
are a standard service-oriented API that defines a broad set of services that are
germane to IT projects in the education field yet also fitting for broader uses.

Sitting alongside of the Harmoni services is the "Harmoni Architecture". The
"architecture" is a set of controllers and templates that coordinate configuration
of services, the proper execution of application actions, and any post-processing of
application output. The architecture is built on a popular module/action model, in
which your PHP program contains multiple modules, each of which contain multiple
executable actions. All you, as an application developer, need to write are the
action files, not the controller logic.

The "Harmoni Architecture" is separate from the Harmoni Services and either can be
used independently of the other.


Current Version Notes
---------------------
This release fixes a minor issue with CAS authentication and visitor support.


Downloads
---------------------
For the latest and archived versions, please download from SourceForge:

http://sourceforge.net/project/showfiles.php?group_id=82873


Documentation
---------------------
See the Harmoni wiki for online documentation:

http://harmoni.sf.net/


Bug Tracker
---------------------
https://sourceforge.net/tracker/?group_id=82873&atid=567473







===================================================================
| Prior Polyphony Release Notes
| (See the Polyphony change log for more details)
===================================================================


v. 1.5.5 (2010-03-04)
----------------------------------------------------
This release fixes a minor issue with CAS authentication and visitor support.



v. 1.5.4 (2009-11-18)
----------------------------------------------------
This release makes a small fix to the authentication support for searching users.



v. 1.5.3 (2009-10-09)
----------------------------------------------------
This release makes a small fix to the Agent support and adds needed search action
and javascript dependencies required by Harmoni-1.10.0's reworked admin-act-as-user
authentication. 



v. 1.5.2 (2009-02-25)
----------------------------------------------------
This release adds a new agent information panel and adds a few features required by
Segue. See the change log for details.



v. 1.5.1 (2009-02-10)
----------------------------------------------------
This release fixes a minor bug in the tag-cloud printing.



v. 1.5.0 (2009-02-09)
----------------------------------------------------
This release includes a number of improvements to tagging and a few other fixes. See
the change-log for details.



v. 1.4.11 (2009-02-02)
----------------------------------------------------
This update adds a minor function to the Javascript Panel's TabbedContent



v. 1.4.10 (2009-01-28)
----------------------------------------------------
This update fixes bugs in the Panel and User-Data Javascript and in the
Group-Membership display. See the change-log for details.



v. 1.4.9 (2009-01-09)
----------------------------------------------------
This release improves some of the group-browsing displays.



v. 1.4.8 (2008-10-15)
----------------------------------------------------
This improves the usability of controls by changing all submit buttons to be labeled
to reflect their action rather than simply saying 'Submit'.



v. 1.4.7 (2008-10-13)
----------------------------------------------------
This release fixes a few small issues. See the change-log for details.



v. 1.4.6 (2008-10-03)
----------------------------------------------------
This release fixes a few small issues relating to tagging, javascript, and file
access. See the change log for details.



v. 1.4.5 (2008-09-24)
----------------------------------------------------
This release fixes a number of small usability issues and adds support for
user-preferences. 



v. 1.4.4 (2008-09-17)
----------------------------------------------------
Thisrelease fixes an issue with file viewing that was causing issues for IE6 users.



v. 1.4.3 (2008-08-27)
----------------------------------------------------
This release add new actions for accessing files from Flash.



v. 1.4.2 (2008-08-25)
----------------------------------------------------
This release includes a few fixes needed by Segue 2.



v. 1.4.1 (2008-08-20)
----------------------------------------------------
This release fixes a minor PHP notice



v. 1.4.0 (2008-08-14)
----------------------------------------------------
This release fixes a number of security issues.

Security issues fixed: 

 * Cross-Site Request Forgeries (CSRF) are now eliminated from data-modification
actions. Read about CSRF at: http://shiflett.org/articles/cross-site-request-forgeries

 * Admin actions are now restricted to prevent listing of users and ids in the
system. 



v. 1.3.3 (2008-08-13)
----------------------------------------------------
This release adds a number of fixes to the help-system to support Segue 2.



v. 1.3.2 (2008-08-07)
----------------------------------------------------
This release fixes several cross-site scripting vulnerabilities in code related to
logging and error display. See the change log for details



v. 1.3.1 (2008-08-01)
----------------------------------------------------
This release fixes a number of Javascript issues that were causing problems in Segue
2. See the change-log for details.



v. 1.3.0 (2008-07-24)
----------------------------------------------------
This release includes several updates to the Javascript library and Wizard to
support Segue 2. See the change-log for details.



v. 1.2.12 (2008-07-22)
----------------------------------------------------
This release fixes an issue that was causing problems in Concerto.



v. 1.2.11 (2008-07-21)
----------------------------------------------------
This release fixes an issue with file uploads under safe mode.



v. 1.2.10 (2008-07-17)
----------------------------------------------------
This release fixes an issue with theme CSS loading that was preventing theme options
from taking effect.



v. 1.2.9 (2008-07-17)
----------------------------------------------------
This release fixes a few bugs to support Segue and Concerto.



v. 1.2.8 (2008-07-14)
----------------------------------------------------
This release fixes a minor issue with log-browsing.



v. 1.2.7 (2008-07-10)
----------------------------------------------------
This release adds a few minor fixes to support Segue.



v. 1.2.6 (2008-06-16)
----------------------------------------------------
This release adds a new multiple checkbox wizard component.



v. 1.2.5 (2008-06-13)
----------------------------------------------------
This release adds some minor improvements.



v. 1.2.4 (2008-06-09)
----------------------------------------------------
This release add new user interfaces used for visitor registration.



v. 1.2.3 (2008-06-03)
----------------------------------------------------
This release fixes a few issues that were affecting Concerto's import and export
functions. 



v. 1.2.2 (2008-05-23)
----------------------------------------------------
Missing theme images now do not fill the logs with errors.



v. 1.2.1 (2008-05-22)
----------------------------------------------------
A parse error had snuck into the thumbnail viewing action. This is now fixed.



v. 1.2.0 (2008-05-20)
----------------------------------------------------
This release adds support for the new Gui2 theming in Harmoni and used by Segue 2.
Also added are a number of new Wizard components and improvements.



v. 1.1.0 (2008-04-11)
----------------------------------------------------
This release add a few minor changes to tagging actions.



v. 1.0.6 (2008-04-03)
----------------------------------------------------
This release adds a minor change needed by Segue 2 - beta 18.



v. 1.0.5 (2008-03-26)
----------------------------------------------------
This release adds a few javascript functions.



v. 1.0.4 (2008-03-25)
----------------------------------------------------
This release fixes a minor syntax error in the wizard.



v. 1.0.3 (2008-03-12)
----------------------------------------------------
Fixes to a few bugs.



v. 1.0.2 (2008-03-10)
----------------------------------------------------
This release fixes a number of bugs, particularly with respect to cross-browser
compatibility and Repository Importer issues.



v. 1.0.1 (2008-02-21)
----------------------------------------------------
This release adds the ability of the Rich-Text editor in the wizard to accept custom
configuration options.



v. 1.0.0 (2008-02-15)
----------------------------------------------------
This release includes a few minor improvements and has been updated to work with
Harmoni 1.0.0.



v. 0.10.7 (2008-01-14)
----------------------------------------------------
New Wizard abilities.



v. 0.10.6 (2007-12-20)
----------------------------------------------------
This release fixes a few minor bugs.



v. 0.10.5 (2007-12-12)
----------------------------------------------------
This release fixes a few minor bugs and adds a new Wizard component.



v. 0.10.4 (2007-11-29)
----------------------------------------------------
This release fixes a number of wizard issues.



v. 0.10.3 (2007-11-09)
----------------------------------------------------
This release fixes a few bugs and adds a new Wizard component.



v. 0.10.2 (2007-11-01)
----------------------------------------------------
This release includes some minor bug fixes and cleanup.



v. 0.10.1 (2007-10-24)
----------------------------------------------------
This release fixes issues with log browsing, admin tools, and tags.



v. 0.10.0 (2007-10-22)
----------------------------------------------------
This release fixes a number of bugs and reworks many of the class hierarchies to use
PHP 5 interfaces and abstract classes for better enforcement of class responsibilities.

The Action class hierarchy has changed to use abstract classes and interfaces. Any
applications that extend the Action class will need to be checked to ensure that
they implement all abstract methods.



v. 0.9.3 (2007-09-25)
----------------------------------------------------




v. 0.9.2 (2007-09-20)
----------------------------------------------------




v. 0.9.1 (2007-09-13)
----------------------------------------------------




v. 0.9.0 (2007-09-07)
----------------------------------------------------




v. 0.8.2 (2007-04-10)
----------------------------------------------------




v. 0.8.1 (2007-04-05)
----------------------------------------------------




v. 0.8.0 (2006-12-13)
----------------------------------------------------




v. 0.7.0 (2006-12-01)
----------------------------------------------------




v. 0.6.9 (2006-11-30)
----------------------------------------------------




v. 0.6.8 (2006-11-28)
----------------------------------------------------




v. 0.6.7 (2006-10-25)
----------------------------------------------------




v. 0.6.6 (2006-08-16)
----------------------------------------------------




v. 0.6.5 (2006-08-15)
----------------------------------------------------




v. 0.6.4 (2006-08-11)
----------------------------------------------------




v. 0.6.3 (2006-08-04)
----------------------------------------------------




v. 0.6.2 (2006-08-02)
----------------------------------------------------




v. 0.6.1 (2006-07-21)
----------------------------------------------------




v. 0.6.0 (2006-06-16)
----------------------------------------------------




v. 0.5.2 (2006-05-26)
----------------------------------------------------




v. 0.5.1 (2006-05-19)
----------------------------------------------------




v. 0.5.0 (2006-05-05)
----------------------------------------------------




v. 0.4.1 (2005-02-09)
----------------------------------------------------




v. 0.4.0 (2005-01-10)
----------------------------------------------------




v. 0.3.0 (2005-10-12)
----------------------------------------------------




v. 0.2.2 (2005-04-14)
----------------------------------------------------




v. 0.2.1 (2005-04-11)
----------------------------------------------------




v. 0.2.0 (2005-04-07)
----------------------------------------------------




v. 0.1.0 (2004-10-26)
----------------------------------------------------






docs/changelog.txt

Polyphony Change Log


v. 1.5.5 (2010-03-04)
----------------------------------------------------
This release fixes a minor issue with CAS authentication and visitor support.
		
		
	* Bug Fix: 
		The visitor login form now doesn't cycle through all authentication methods.
		
		 The CAS AuthNMethod was getting called first and forwarding visitors to
		CAS where they wouldn't be able to log in.
		(2010-03-04 - Adam Franco)
	



v. 1.5.4 (2009-11-18)
----------------------------------------------------
This release makes a small fix to the authentication support for searching users.
		
		
	* Bug Fix: 
		Change-user results are now printed one per line. Fixes issue:
		http://chisel.middlebury.edu/pool/show_bug.cgi?id=353 
		(2009-10-28 - Adam Franco)
	



v. 1.5.3 (2009-10-09)
----------------------------------------------------
This release makes a small fix to the Agent support and adds needed search action
and javascript dependencies required by Harmoni-1.10.0's reworked admin-act-as-user
authentication. 
		
		
	* Bug Fix: 
		Agent: Group brows UI now doesn't die if group properties aren't available.
		(2009-10-09 - Adam Franco)
		
	* New feature: 
		Authentication: Added search_users action and jquery dependencies for new
		Admin-Act-As-User authentication.
		(2009-10-09 - Adam Franco)
	



v. 1.5.2 (2009-02-25)
----------------------------------------------------
This release adds a new agent information panel and adds a few features required by
Segue. See the change log for details.
		
		
	* New feature: 
		Agent: New Agent-Info panel for displaying details about an agent.
		(2009-02-17 - Adam Franco)
		
	* Change: 
		Wizard: The RowRadioMatrix now supports embedding of components in its
		descriptions. 
		(2009-02-17 - Adam Franco)
		
	* New feature: 
		Javascript: New utility method document.getChildrenByTagName(target,
		tagName). 
		(2009-02-17 - Adam Franco)
		
	* New feature: 
		Authentication: New PolyphonyLogin class to allow client apps to assess if a
		login attempt has just failed.
		(2009-02-23 - Adam Franco)
	



v. 1.5.1 (2009-02-10)
----------------------------------------------------
This release fixes a minor bug in the tag-cloud printing.
		
		
	* Bug Fix: 
		Tagging: Tag grouping for the cloud will now not exceed the number of styles
		defined. 
		(2009-02-10 - Adam Franco)
	



v. 1.5.0 (2009-02-09)
----------------------------------------------------
This release includes a number of improvements to tagging and a few other fixes. See
the change-log for details.
		
		
	* New feature: 
		Tagging: Can now display a list of tags in addition to the cloud.
		(2009-02-09 - Adam Franco)
		
	* Bug Fix: 
		Tagging: TagCloud writing now can take in a 2D array of name-spaced
		parameters to pass through to the links in the tag cloud.
		(2009-02-09 - Adam Franco)
		
	* Change: 
		Javascript: document.get_elements_by_class() now accepts an optional root
		element for the search.
		(2009-02-09 - Adam Franco)
		
	* Change: 
		Actions: Can now support a page-title different from the heading.
		(2009-02-09 - Adam Franco)
	



v. 1.4.11 (2009-02-02)
----------------------------------------------------
This update adds a minor function to the Javascript Panel's TabbedContent
		
		
	* New feature: 
		Javascript TabbedContent now supports a getSelected() function.
		(2009-02-02 - Adam Franco)
	



v. 1.4.10 (2009-01-28)
----------------------------------------------------
This update fixes bugs in the Panel and User-Data Javascript and in the
Group-Membership display. See the change-log for details.
		
		
	* Change: 
		UserData: New access method for returning UserData javascript. The
		writeHeadJs() method now only writes the javascript once.
		(2009-01-14 - Adam Franco)
		
	* Bug Fix: 
		Agents: Group-membership screen now doesn't crash on groups that do not
		support properties fetching.
		(2009-01-27 - Adam Franco)
		
	* Bug Fix: 
		Javascript: CenteredPanel.moveToFront() now sets the zIndex of the screen as
		well. 
		(2009-01-28 - Adam Franco)
	



v. 1.4.9 (2009-01-09)
----------------------------------------------------
This release improves some of the group-browsing displays.
		
		
	* Bug Fix: 
		AuthZ: Choose-Agent UI now includes group descriptions as well as type
		descriptions. 
		(2008-10-20 - Adam Franco)
	



v. 1.4.8 (2008-10-15)
----------------------------------------------------
This improves the usability of controls by changing all submit buttons to be labeled
to reflect their action rather than simply saying 'Submit'.
		
		
	* Bug Fix: #2033088 
		All submit buttons have been labeled to reflect their action rather than
		simply saying 'Submit'.
		(2008-10-15 - Adam Franco)
		(reported by Alex Chapin of Middlebury College)
	



v. 1.4.7 (2008-10-13)
----------------------------------------------------
This release fixes a few small issues. See the change-log for details.
		
		
	* Bug Fix: 
		Repository: View-File action now uses the first file record if no record-id
		is specified rather than throwing an exception.
		(2008-10-06 - Adam Franco)
		
	* Bug Fix: 
		Logs: Browsing of logs is now properly restricted to administrators.
		(2008-10-10 - Adam Franco)
	



v. 1.4.6 (2008-10-03)
----------------------------------------------------
This release fixes a few small issues relating to tagging, javascript, and file
access. See the change log for details.
		
		
	* Bug Fix: #1939677 
		Tagging: Tag clouds written by Javascript now include passed-through
		parameters. 
		(2008-09-30 - Adam Franco)
		
	* Change: #1939677 
		Javascript: Harmoni.quickURL() now supports parameters with mixed namespaces.
		
		Normal parameters are still specified as: {key1: 'value1', key2: 'value2'}
		
		Rather than specifying a string value, an object with 'value' and
		'namespace' properties can be given: {key1: 'value1', key2: {value:
		'value2', namespace: 'foo'}}
		(2008-09-30 - Adam Franco)
		
	* Change: #2138652 

javascript/Panel.js
/**
 * @since 11/29/06
 * @package polyphony.javascript
 * 
 * @copyright Copyright &copy; 2005, Middlebury College
 * @license http://www.gnu.org/copyleft/gpl.html GNU General Public License (GPL)
 *
 * @version $Id: Panel.js,v 1.5 2007/05/03 20:56:12 adamfranco Exp $
 */

/**
 * The panel is an absolutely positioned element that can be positioned relative
 * to a calling element. The panel has a title-bar with close button and a content
 * area. The panel caches itself on the page and can be closed and reopened without
 * loosing its state.
 * 
 * @since 11/29/06
 * @package polyphony.javascript
 * 
 * @copyright Copyright &copy; 2005, Middlebury College
 * @license http://www.gnu.org/copyleft/gpl.html GNU General Public License (GPL)
 *
 * @version $Id: Panel.js,v 1.5 2007/05/03 20:56:12 adamfranco Exp $
 */
function Panel ( title, height, width, positionElement, classNames ) {
	if ( arguments.length > 0 ) {
		this.init( title, height, width, positionElement, classNames);
	}
}

	/**
	 * Initialize this panel
	 * 
	 * @param string 	title
	 * @param integer	height
	 * @param integer	width
	 * @param optional DOM_Element positionElement
	 *		A unique element that this panel is associated with. An element can 
	 *		only have one panel associated with it, which will be cached with 
	 *		this element.
	 * @param string	classNames	Names of CSS classes to apply to the panel.
	 * @return void
	 * @access public
	 * @since 11/29/06
	 */
	Panel.prototype.init = function ( title, height, width, positionElement, 
		classNames ) 
	{
		if (positionElement) {
			this.positionElement = positionElement;
			this.positionElement.panel = this;
		}
		
		this.title = title;
		this.height = height;
		this.width = width;
		this.classNames = classNames;
		this.createElements();
		
	}
	
	/**
	 * Open or create the panel as needed
	 * 
	 * @return object Panel
	 * @access public
	 * @since 11/29/06
	 * @static
	 */
	Panel.run = function ( title, height, width, positionElement, classNames ) {
		if (positionElement.panel) {
			var panel = positionElement.panel;
			panel.open();
			
		} else {
			var panel = new Panel( title, height, width, positionElement, classNames );
		}
		
		if (panel.onOpen) {
			panel.onOpen();
		}
		
		return panel;
	}
	
	/**
	 * Create the panel elements
	 * 
	 * @return void
	 * @access public
	 * @since 11/29/06
	 */
	Panel.prototype.createElements = function () {
		this.mainElement = document.createElement("div");
		this.mainElement.panel = this;
		if (this.classNames)
			this.mainElement.className = 'panel ' + this.classNames;
		else
			this.mainElement.className = 'panel';
		
		this.positionMainElement();
		
		// Top bar
		this.topBar = this.mainElement.appendChild(document.createElement("table"));
		this.topBar.className = 'topbar';
		this.topBar.style.width = '100%';

		var tbody = this.topBar.appendChild(document.createElement('tbody'));
		var row1 = tbody.appendChild(document.createElement('tr'));
		
		var topLeft = row1.appendChild(document.createElement('td'));
		topLeft.className = 'title';
		topLeft.innerHTML = this.title;
		this.titleElement = topLeft;
		
		var topRight = row1.appendChild(document.createElement('td'));
		topRight.className = 'close';
		this.cancel = topRight.appendChild(document.createElement("a"));
		var panel = this;	// define a variable for panel that will be in the
							// scope of the onclick.
		this.cancel.onclick = function () {panel.close();}
		this.cancel.innerHTML = 'Close';
		
		// Content container
		this.contentElement = document.createElement("div");
		this.mainElement.appendChild(this.contentElement);
		this.contentElement.className = 'container';
		
		this.moveToFront();
		document.body.appendChild(this.mainElement);
	}
	
	/**
	 * Position the main element
	 * 
	 * @return void
	 * @access public
	 * @since 7/31/08
	 */
	Panel.prototype.positionMainElement = function () {
// 		this.mainElement.style.height = this.height + 'px';
		this.mainElement.style.width = this.width + 'px';
		this.mainElement.style.position = 'absolute';
		this.mainElement.style.overflow = 'auto';
		this.mainElement.style.top = this.getTop() + "px";
		this.mainElement.style.left = this.getLeft() + "px";
	}
	
	/**
	 * Give the panel a z-index greater than any other item.
	 * 
	 * @return void
	 * @access public
	 * @since 9/24/08
	 */
	Panel.prototype.moveToFront = function () {
		var elements = document.getElementsByTagName("*");
		var maxZIndex = 0;
		for (var i = 0; i < elements.length; i++) {
			if (elements[i] !== this.mainElement) {
				maxZIndex = Math.max(elements[i].style.zIndex, maxZIndex);
			}
		}
		
		this.mainElement.style.zIndex = maxZIndex + 1;
	}
	
	/**
	 * Answer the number of pixels from the top of the screen to position the
	 * panel.
	 * 
	 * @return integer
	 * @access public
	 * @since 1/26/07
	 */
	Panel.prototype.getTop = function () {
		var top = (document.getOffsetTop(this.positionElement) 
						- Math.round(this.height / 2) 
						+ Math.round(this.positionElement.offsetHeight / 2));
		if (top < 5)
			top = 5;
		
		return top;
	}
	
	/**
	 * Answer the number of pixels from the left of the screen to position the
	 * panel.
	 * 
	 * @return integer
	 * @access public
	 * @since 1/26/07
	 */
	Panel.prototype.getLeft = function () {
		var left = (document.getOffsetLeft(this.positionElement) 
						- Math.round(this.width / 2) 
						+ Math.round(this.positionElement.offsetWidth / 2));
		if (left < 5)
			left = 5;
		
		return left;
	}
	
	/**
	 * Close the panel. Override the onClose() method of this object to add other
	 * actions to do on close.
	 * 
	 * @return void
	 * @access public
	 * @since 11/29/06
	 */
	Panel.prototype.close = function () {
		this.mainElement.style.display = 'none';
		if (this.onClose) {
			this.onClose();
		}
	}
	
	/**
	 * Open the panel. Override the onOpen() method of this object to add other
	 * actions to do on open.
	 * 
	 * @return void
	 * @access public
	 * @since 11/29/06
	 */
	Panel.prototype.open = function () {
		this.mainElement.style.display = 'block';
		this.moveToFront();
		if (this.onOpen) {
			this.onOpen();
		}
	}
	
	/**
	 * Reposition the main element based on the rendered height
	 * 
	 * @return void
	 * @access public
	 * @since 1/26/07
	 */
	Panel.prototype.centerOnHeight = function () {
		this.height = this.mainElement.offsetHeight;		
		this.mainElement.style.top = this.getTop() + "px";
	}
	
	/**
	 * Reposition the main element based on the rendered width
	 * 
	 * @return void
	 * @access public
	 * @since 1/26/07
	 */
	Panel.prototype.centerOnWidth = function () {
		this.width = this.mainElement.offsetWidth;
		this.mainElement.style.left = this.getLeft() + "px";
	}
	
	/**
	 * Reposition the main element based on the rendered height and width
	 * 
	 * @return void
	 * @access public
	 * @since 1/26/07
	 */
	Panel.prototype.center = function () {
		this.centerOnHeight();
		this.centerOnWidth();
	}